FLT/HIMExperiments/module_topology.lean

```diff
@@ -137,13 +136,81 @@ example (ι : Type*) [Finite ι] :
 
+--maybe I should conclude this from the example above but don't know how
+--otherwise this should be very easy to do by hand
+lemma Module.topology_self : (hA : TopologicalSpace A) = Module.topology A := by
+  apply le_antisymm
+  · unfold topology
+    apply le_iInf
+    intro i
+    rw [← continuous_iff_le_induced]
+    have hi : i = LinearMap.lsmul A A (i 1) := by
+      ext
+      simp only [LinearMap.lsmul_apply, smul_eq_mul, mul_one]
+    rw [hi]
+    apply continuous_const.mul continuous_id
+    -- a linear map should always be continuous in this setting right?
+  · apply sInf_le
+    rw [Set.mem_range]
+    use LinearMap.id
+    exact induced_id
+
 -- We need that the module topology on a product is the product topology
 lemma Module.prod_canonical :
     @instTopologicalSpaceProd M N (Module.topology A) (Module.topology A) =
     (Module.topology A : TopologicalSpace (M × N)) := by
   -- the goal is to prove that an iInf equals an inf (of two topologies).
-  unfold instTopologicalSpaceProd -- you can probably delete this later
-  sorry
+  apply le_antisymm
+  · apply le_iInf
+    intro f
+    rw [← continuous_iff_le_induced]
+    have : ⇑f = fun ⟨m, n⟩ ↦ f (⟨m, 0⟩) + f (⟨0, n⟩) := by
+      ext x
+      simp only
+      have : x = (x.1, 0) + (0, x.2) := by simp
+      nth_rewrite 1 [this]
+      apply LinearMap.map_add
+    rw [this]
+    simp only
+    have h1 : @Continuous _ _ (@instTopologicalSpaceProd M N (Module.topology A) (Module.topology A))
+      _ (fun (x : M × N) ↦ f (x.1, 0)) := by
+        have : (fun (x : M × N) ↦ f (x.1, 0)) = fun x ↦ (fun m ↦ f (m, 0)) x.1 := by
+          ext x
+          simp only [Function.comp_apply]
+        rw [this]
+        apply @Continuous.fst' _ _ _ (Module.topology A) (Module.topology A) _ (fun m ↦ f (m, 0))
+        let h :  M →ₗ[A] A := {
+          toFun := fun m ↦ f (m, 0)
+          map_add' := by
+            intro x y
+            rw [← LinearMap.map_add, Prod.mk_add_mk, add_zero]
+          map_smul' := by
+            intro m x
+            rw [← LinearMap.map_smul, RingHom.id_apply, Prod.smul_mk, smul_zero]
+        }
+        have : (fun m ↦ f (m, 0)) = h := by rfl
+        rw [this, continuous_iff_le_induced]
+        nth_rewrite 2 [Module.topology_self A]
+        rw [← continuous_iff_le_induced]
+        exact Module.continuous_linear A h
+    have h2 : @Continuous _ _ (@instTopologicalSpaceProd M N (Module.topology A) (Module.topology A))
+      _ (fun (x : M × N) ↦ f (0, x.2))
+      := by sorry -- this will be exactly the same as above so better do the cleanup first
+    apply @Continuous.add _ _
+      (@instTopologicalSpaceProd M N (Module.topology A) (Module.topology A)) _ _ _ _ _ h1 h2
+  · apply le_inf
+    · rw [induced_iInf]
+      apply le_iInf
+      intro f
+      rw [induced_compose]
+      exact iInf_le (fun (f :  M × N →ₗ[A] A) ↦ TopologicalSpace.induced (⇑f) inferInstance)
+        (LinearMap.lcomp A A (LinearMap.fst A M N) f)
```

Comment:
```suggestion
      exact iInf_le _ (LinearMap.lcomp _ _ (LinearMap.fst _ _ _) _)
```
I can imagine that you may well have filled in a lot of those holes when you were golfing this proof, but I'm just observing that if you want you can drop many of them, because `exact` is good at unification.